ipn/ipnauth/access.go

```diff
@@ -0,0 +1,240 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package ipnauth
+
+import (
+	"math/bits"
```

Comment:
Having access rights as bit masks serves as both a space and time optimization, and it makes them easier to work with. For instance, it simplifies expressing multiple access rights simultaneously when an operation requires more than one. Additionally, I think this approach makes both the calling code and the actual access checking implementation less verbose and easier to read.

While it does make stringification more complex than alternative approaches, it seems to be a reasonable trade-off. Also, FWIW, in most cases, there's no need to convert `DeviceAccess` and `ProfileAccess` into their string representations. In fact, we primarily use this for error formatting in unit tests.

Reply:
We can agree to disagree :) but IMO this looks like code that makes a bit more sense in C (which has real enums and bitfields and lacks tailscale.com/util/Set and map) but which reads kind of clunkily in Go. If space/time the optimizations are necessary, I understand; but I prefer to optimize for Obviously Correct in places like access control logic. Given, e.g.,

```
type DeviceAccess Set[bool]
const (
    ReadDeviceStatus = true
    // ...
)
```

you can have call sites like

```
access.Add(ReadDeviceStatus)
// ...
access.Delete(OtherStuff)
// ...

if access.Contains(ReadDeviceStatus) && access.Contains(Debug) {
    // ...
}
```

which is Obvious (at least, to me) the way bitwise arithmetic sometimes isn't (at least, to me). For example, with bitmasks, people sometimes do things like

```
const (
	Foo int = 1 << iota
	Bar
)

func main() {
	accessA := 3
	accessB := 1
	if accessA&(Foo|Bar) != 0 {
		fmt.Println("expected")
	}
	if accessB&(Foo|Bar) != 0 {
		fmt.Println("oops, we reach this branch and that is a bug — should be `accessB&(Foo|Bar) == Foo|Bar")
	}
}
```

(And that's before we get into `&` vs. `&&` typos, et c.)

If we have evidence that we do need the time and/or space optimization of bitfields, we might consider wrapping such an implementation in a readable, Go-idiomatic API. (The convenience functions will just be bitmasking, and will most likely get compiled inline.)

Reply:
> we might consider wrapping such an implementation in a readable, Go-idiomatic API

Code in the `ipnauth` package _is_ the wrapper. At call sites, I was considering having code like this:
```
if ipnauth.CheckProfileAccess(caller, profile, prefs, ipnauth.ReadProfileInfo).Denied() {
    continue
}
```
and sometimes this:
```
requiredAccess := ipnauth.Connect | ipnauth.ReauthProfile
if err := ipnauth.CheckProfileAccess(caller, profile, prefs, requiredAccess).Error(); err != nil {
    return err
}
```
In my opinion, this is quite easy to read, review and reason about:
> requiredAccess := ipnauth.Connect | ipnauth.ReauthProfile

<br/>

But if we prefer, we can make make a wrapper and use it somewhat like this:
```
var requiredAccess ipnauth.ProfileAccess
requiredAccess.AddReadProfileInfo()
if err := ipnauth.CheckProfileAccess(caller, profile, prefs, requiredAccess).Error(); err != nil {
    return ipn.LoginProfile{}, err
}
```
Alternatively, we could add a number of convenience methods like the following:
```
func (da *DeviceAccess) Add(a DeviceAccess) {
	*da |= a
}

func (da *DeviceAccess) Remove(a DeviceAccess) {
	*da &= ^a
}
```
and use them at both external call sights and within `ipnauth`. In this case, we still benefit from the time and space optimizations and can keep code like `ipnauth.CheckProfileAccess(caller, profile, prefs, ipnauth.ReadProfileInfo)` as is. However, this approach also allows us to express our intentions more clearly by using the `Add` and `Remove` methods instead of directly using bitwise operations.

What do you think?

Reply:
Ahh, ok. I see. That seems reasonable to me. I'd rather have only 1 of Add/Remove vs. bitwise ops. (There's certainly precedent in Go for bitwise |-ing public constants, e.g. in `log`.) Whichever you prefer is fine by me. :)